crates/cli-support/src/js/mod.rs

```diff
@@ -1344,18 +1344,51 @@ impl<'a> Context<'a> {
         self.expose_text_encoder();
         self.expose_uint8_memory();
 
+        // A fast path that directly writes char codes into WASM memory as long
+        // as it finds only ASCII characters.
+        //
+        // This is much faster for common ASCII strings because it can avoid
+        // calling out into C++ TextEncoder code.
+        //
+        // This might be not very intuitive, but such calls are usually more
+        // expensive in mainstream engines than staying in the JS, and
+        // charCodeAt on ASCII strings is usually optimised to raw bytes.
+        let start_encoding_as_ascii = format!(
+            "
+                {}
+                let size = arg.length;
+                let ptr = wasm.__wbindgen_malloc(size);
+                let offset = 0;
+                {{
+                    const mem = getUint8Memory();
+                    for (; offset < arg.length; offset++) {{
+                        const code = arg.charCodeAt(offset);
+                        if (code > 0x7F) {{
+                            arg = arg.slice(offset);
+                            break;
+                        }}
+                        mem[ptr + offset] = code;
+                    }}
+                }}
+            ",
+            debug
+        );
+
         // The first implementation we have for this is to use
         // `TextEncoder#encode` which has been around for quite some time.
         let use_encode = format!(
             "
                 {}
-                const buf = cachedTextEncoder.encode(arg);
-                const ptr = wasm.__wbindgen_malloc(buf.length);
-                getUint8Memory().set(buf, ptr);
-                WASM_VECTOR_LEN = buf.length;
+                if (offset !== arg.length) {{
+                    const buf = cachedTextEncoder.encode(arg);
+                    ptr = wasm.__wbindgen_realloc(ptr, size, size += buf.length);
```

Comment:
I think that this may be reallocating to too much memory? `size += buf.length` is `arg.length + encodedSize(arg_substring_startin_with_non_ascii)` but I think that this should instead reset `size` to `offset + buf.length`?

Reply:
Ah, you are probably right.

So for example, if you had a string `"abc😊"`, then by now `size == 5`, `offset == 3`, `arg == "😊"`, `buf == [ 240, 159, 152, 138 ]` and we'll get `size + buf.length == 9` but we need only `offset + buf.length == 7`.

Reply:
Fixed.